src/coreclr/vm/frozenobjectheap.cpp

```diff
@@ -0,0 +1,162 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include "frozenobjectheap.h"
+#include "memorypool.h"
+
+
+FrozenObjectHeapManager::FrozenObjectHeapManager():
+    m_CurrentHeap(nullptr)
+{
+    m_Crst.Init(CrstFrozenObjectHeap, CRST_UNSAFE_COOPGC);
+    m_HeapCommitChunkSize = CLRConfig::GetConfigValue(CLRConfig::INTERNAL_FrozenSegmentCommitSize);
+    m_HeapSize = CLRConfig::GetConfigValue(CLRConfig::INTERNAL_FrozenSegmentReserveSize);
+}
+
+// Allocates an object of the give size (including header) on a frozen segment.
+// May return nullptr in the following cases:
+//   1) DOTNET_FrozenSegmentReserveSize is 0 (disabled)
+//   2) Object is too large (large than DOTNET_FrozenSegmentCommitSize)
+// in such cases caller is responsible to find a more appropriate heap to allocate it
+Object* FrozenObjectHeapManager::AllocateObject(PTR_MethodTable type, size_t objectSize)
+{
+    CONTRACTL
+    {
+        THROWS;
+        MODE_COOPERATIVE;
+    }
+    CONTRACTL_END
+
+    CrstHolder ch(&m_Crst);
+
+    // Quick way to disable Frozen Heaps
+    if (m_HeapSize == 0)
+    {
+        return nullptr;
+    }
+
+    _ASSERT(type != nullptr);
+    _ASSERT(m_HeapCommitChunkSize >= MIN_OBJECT_SIZE);
+    _ASSERT(m_HeapSize > m_HeapCommitChunkSize);
+    _ASSERT(m_HeapSize % m_HeapCommitChunkSize == 0);
+
+    // NOTE: objectSize is expected be the full size including header
+    _ASSERT(objectSize >= MIN_OBJECT_SIZE);
+
+    if (objectSize > m_HeapCommitChunkSize)
+    {
+        // The current design doesn't allow object larger than DOTNET_FrozenSegmentCommitSize and
+        // since FrozenObjectHeap is just an optimization, let's not fill it with huge objects.
+        return nullptr;
+    }
```

Comment:
I do not think that this special-case makes sense. I would allocate all interned strings on the frozen heap, no exceptions.

Reply:
Actually, there seems to be other situations that de-rail this optimization. For example, if a collectible assembly interns the string first, the non-collectible assembly that comes after it won't be able to take advantage of this optimizations. Given that, I think that skipping the optimization for large strings is ok.